File: native/ConceprFirstRxSwift/rxProxy/ObservableImpl.swift

import Foundation
import SharedCode
import RxSwift

class ObservableImpl<T> : SharedCode.Observable {
    
    init(observable: RxSwift.Observable<T>) {
      self.observable = observable
    }

    var observable: RxSwift.Observable<T>
    
    func subscribe(onNext: @escaping (Any?) -> Void, onError: @escaping (KotlinThrowable) -> Void) -> SharedCode.Disposable {
           
           return DisposableImpl(disposable:
            self.observable.subscribe { event in
                
                switch event {
                    
                case .next(let value):
                    onNext(value)
                case .error(let error):
                    onError(self.convertToKotlinThrowable(error: error))
                default:
                    break
                }
            }
           )
       }
    
    func subscribeOn(scheduler: Scheduler) -> SharedCode.Observable {
        return self
    }
    
    func observeOn(scheduler: Scheduler) -> SharedCode.Observable {
        return self
    }
    
    func delay(millisec: Int64) -> SharedCode.Observable {
        return observable.delay(.milliseconds(millisec), null)
    }
    
    private func convertToKotlinThrowable (error: Error)-> KotlinThrowable {
        
        return KotlinThrowable(message: "error from rx swift")
    }
    
}

File: SharedCode/src/androidMain/kotlin/rxProxy/ObservableImpl.kt
package com.jetbrains.handson.mpp.mobile.rxProxy

import com.jetbrains.handson.mpp.mobile.rxProxy.Scheduler
import com.jetbrains.handson.mpp.mobile.the_rx.Disposable
import com.jetbrains.handson.mpp.mobile.the_rx.Observable
import java.util.concurrent.TimeUnit

class ObservableImpl<T>(
    private var observable: io.reactivex.Observable<T>
): Observable<T> {

    override fun subscribe(onNext: (T) -> Unit, onError: (Throwable) -> Unit): Disposable {
        return DisposableImpl(
            observable.subscribe(
                onNext,
                onError
            )
        )
    }

    override fun subscribeOn(scheduler: Scheduler): Observable<T> {
        return ObservableImpl(observable.subscribeOn((scheduler as SchedulerImpl).scheduler))
    }

    override fun observeOn(scheduler: Scheduler): Observable<T> {
        return ObservableImpl(observable.observeOn((scheduler as SchedulerImpl).scheduler))
    }

    override fun delay(millisec: Long): Observable<T> {
        return ObservableImpl(
            observable.delay(
                millisec,
                TimeUnit.MILLISECONDS
            )
        )
    }
}